src/main/java/games/strategy/engine/ClientFileSystemHelper.java

```diff
@@ -30,66 +29,51 @@ private ClientFileSystemHelper() {}
-    final URL url = GameRunner.class.getResource(runnerClassName);
-    final String path = url.getFile();
     try {
-      // Deal with spaces in the file name which would be url encoded
-      return URLDecoder.decode(path, StandardCharsets.UTF_8.name());
-    } catch (final UnsupportedEncodingException e) {
-      throw new AssertionError("platform does not support UTF-8 charset", e);
+      return getFolderContainingFileWithName(getCodeSourceFolder(), GAME_ENGINE_PROPERTIES_FILE);
+    } catch (final IOException e) {
+      throw new IllegalStateException("unable to locate root folder", e);
     }
   }
 
-  private static File getRootFolderRelativeToJar(final String path, final String jarFileName) {
-    final String subString = path.substring(
-        "file:/".length() - (SystemProperties.isWindows() ? 0 : 1),
-        path.indexOf(jarFileName) - 1);
-    final File f = new File(subString).getParentFile();
-    if (!f.exists()) {
-      throw new IllegalStateException("File not found:" + f);
+  private static File getCodeSourceFolder() throws IOException {
+    final @Nullable CodeSource codeSource = ClientFileSystemHelper.class.getProtectionDomain().getCodeSource();
+    if (codeSource == null) {
+      throw new IOException("code source is not available");
     }
-    return f;
-  }
 
-  private static File getRootFolderRelativeToClassFile(final String path) {
-    File f = new File(path);
-
-    // move up one directory for each package
-    final int moveUpCount = GameRunner.class.getName().split("\\.").length + 1;
-    for (int i = 0; i < moveUpCount; i++) {
-      f = f.getParentFile();
+    final File codeSourceLocation;
+    try {
+      codeSourceLocation = new File(codeSource.getLocation().toURI());
+    } catch (final URISyntaxException e) {
+      throw new IOException("code source location URI is malformed", e);
     }
 
-    // keep moving up one directory until we find the game_engine properties file that we expect to be at the root
-    while (!folderContainsGamePropsFile(f)) {
-      f = f.getParentFile();
-    }
+    // code source location is either a jar file (installation) or a folder (dev environment)
+    return codeSourceLocation.isFile() ? codeSourceLocation.getParentFile() : codeSourceLocation;
+  }
 
-    if (!f.exists()) {
-      System.err.println("Could not find root folder, does  not exist:" + f);
-      return new File(SystemProperties.getUserDir());
+  @VisibleForTesting
+  static File getFolderContainingFileWithName(final File startFolder, final String fileName) throws IOException {
+    @Nullable
+    File folder;
+    for (folder = startFolder; folder != null; folder = folder.getParentFile()) {
+      final boolean containsFileWithName = FileUtils.listFiles(folder).stream()
+          .filter(File::isFile)
+          .map(File::getName)
+          .anyMatch(fileName::equals);
+      if (containsFileWithName) {
+        return folder;
+      }
     }
```

Comment:
Something bugs me about this iterative implementation.
A recursive implementation would be much more fitting for this kind of purpose.

Reply:
Something like this:
```diff
diff --git a/src/main/java/games/strategy/engine/ClientFileSystemHelper.java b/src/main/java/games/strategy/engine/ClientFileSystemHelper.java
index 43e5525..5642f28 100644
--- a/src/main/java/games/strategy/engine/ClientFileSystemHelper.java
+++ b/src/main/java/games/strategy/engine/ClientFileSystemHelper.java
@@ -58,22 +58,24 @@ public final class ClientFileSystemHelper {
   }

   @VisibleForTesting
-  static File getFolderContainingFileWithName(final File startFolder, final String fileName) throws IOException {
-    @Nullable
-    File folder;
-    for (folder = startFolder; folder != null; folder = folder.getParentFile()) {
-      final boolean containsFileWithName = FileUtils.listFiles(folder).stream()
-          .filter(File::isFile)
-          .map(File::getName)
-          .anyMatch(fileName::equals);
-      if (containsFileWithName) {
-        return folder;
-      }
-    }
+  static File getFolderContainingFileWithName(final File folder, final String fileName) throws IOException {
+    return getFolderContainingFileWithName(folder, folder, fileName);
+  }

-    throw new IOException(String.format(
-        "unable to locate file with name '%s' starting from folder '%s'",
-        fileName, startFolder.getAbsolutePath()));
+  private static File getFolderContainingFileWithName(final File startFolder, final File folder, final String fileName)
+      throws IOException {
+    if (folder == null) {
+      throw new IOException(String.format(
+          "unable to locate file with name '%s' starting from folder '%s'",
+          fileName, startFolder.getAbsolutePath()));
+    }
+    final boolean containsFileWithName = FileUtils.listFiles(folder).stream()
+        .filter(File::isFile)
+        .map(File::getName)
+        .anyMatch(fileName::equals);
+    return containsFileWithName
+        ? folder
+        : getFolderContainingFileWithName(startFolder, folder.getParentFile(), fileName);
   }

   /**
```
(Can be simplyfied to a single method, if the error message doesn't need to know about the original start folder)

Reply:
Yeah, I didn't consider recursion, but I did consider making the `startFolder` parameter mutable so I could just reuse it in the method body in lieu of `folder`.  But I left the implementation as-is because of the desire to display a meaningful path in the error message.  Just as in the recursive case, by the time you get to the terminating condition, the folder variable is `null` and you can't really use it in the error message. 😞 

Although I think the recursive solution may be a bit much here, I see it having two fewer smells than the iterative solution: no Eclipse formatter kludge, and no mutable variables.  So I'll make the change as recommended.  Thanks.

Reply:
2e79793e5